### ISSUE TITLE
[no-unused-vars] (v4-alpha.1) Don't warn about unused type params (or add an option)

### ISSUE THREAD

**Issue by eamodio** [typescript-eslint/typescript-eslint#2337]
While this is probably not a bug, I don't know if I would consider type parameters in a rule about unused vars. Or at least I would prefer to have a way to opt out of them being included.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": [
		"warn",
		{
			"args": "after-used",
			"argsIgnorePattern": "^_",
			"ignoreRestSiblings": true,
			"varsIgnorePattern": "^_$"
		}
	]
  }
}
```

```TS
export type IpcCommandParamsOf<CT> = CT extends IpcCommandType<infer P> ? P : never;
export class IpcCommandType<P = any> {
	constructor(public readonly method: string) {}
}
```

In that example, while the `IpcCommandType` class itself doesn't use the `P` type parameter, it is information carried along with the class for better type inference/safety in certain usages. I use the `IpcCommandParamsOf` type to extract that `P` type from the `IpcCommandType` class. Which can be used in something like the following function:

```ts
export function onIpcCommand<CT extends IpcCommandType>(
	type: CT,
	command: IpcMessage,
	fn: (params: IpcCommandParamsOf<CT>) => unknown,
) {
	fn(command.params);
}
```

**Expected Result**

No errors

**Actual Result**

```
17:29  warning  'P' is defined but never used. Allowed unused vars must match /^_$/u  @typescript-eslint/no-unused-vars
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.0.0-alpha.1` |
| `@typescript-eslint/parser`        | `4.0.0-alpha.1` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.5.0` |
| `node`                             | `12.16.2` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
THat's an interesting pattern - I've never seen it before.
Could you prefix the type with `_` so that it is ignored?

I think it makes sense to add a `typeParameters` option similar to the other options like `vars`, `args`, and `caughtErrors`.
Likely makes sense to also add a `typeDeclaration` option as well.

For both also would make sense to add a `*IgnorePattern` option as well.

**Comment by eamodio** [comment]
I tried the `_` prefix, but it still seems to cause the same warning:
```
17:29  warning  '_P' is defined but never used. Allowed unused vars must match /^_$/u  @typescript-eslint/no-unused-vars
```

**Comment by dbartholomae** [comment]
Wouldn't this generic be stripped anyways by TypeScript? See https://github.com/Microsoft/TypeScript/wiki/FAQ#why-is-astring-assignable-to-anumber-for-interface-at--

**Comment by bradzacher** [comment]
I think the important part is this:

> I use the IpcCommandParamsOf type to extract that P type from the IpcCommandType class.

Even if the type is unused, its information is carried along, and you can then extract it later:

```ts
type Test = IpcCommandParamsOf<IpcCommandType<string>>; // T === string
```

**Comment by tadhgmister** [comment]
as of the latest version:
```ts
export class Test<_P> {}
```
This no longer complains that `_P` is unused so is this issue closed?

**Comment by bradzacher** [comment]
I think I'm okay with closing this - I don't think it's worth adding a separate option purely for types or type parameters if `varsIgnorePattern` covers it.

**Comment by mikaelpaj** [comment]
Phantom type parameters can be quite useful, though I suppose they aren't used that much in TS. At least not yet. We use them for type safe requests like this (simplified):

```ts
type Product = { name: string /* ... */ }
type Route<Result, Path> = { path: Path }

const productsRoute: Route<Product[], '/products'> = { path: '/products' }

type Fetch<Result> = (route: Route<Result, string>) => Result
```

Something like the `typeParameters` option would be a better solution for us than having to switch to a specific naming pattern and using `varsIgnorePattern`.

**Comment by bradzacher** [comment]
There wouldn't be an option to ignore type parameters altogether. There would only be an option like `typeParametersIgnorePattern`.

Which is why I say I don't think it's really worth it.

Most codebase I've seen configure `varsIgnorePattern` and `argsIgnorePattern` to be the same thing (which is usually `^_`). No doubt most codebases would just add a 3rd config with the same pattern.

This is why I say it's probably not worth it. This also creates a weird behaviour in that types aren't ignorable by it, but type parameters are. So then do you add one for types as well?

Without a strong requirement for a separate option, I think for now it's fine having one option.

Also adding a second option here is technically a breaking change.
